# 0. Environment

### Import Modules

In [451]:
import pandas as pd
import numpy as np
import pandas_profiling
import re
pd.set_option('display.max_columns',None)
pd.set_option('display.max_colwidth', 200)

### Import CSV

In [452]:
data = pd.read_csv('GSAF5.csv',encoding='iso8859-1')

# 1. Checking the DataFrame 

## 1.1 General look of the DataFrame

In [453]:
data.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   Age  \
0  New Smyrna Beach, Volusia County  Surfing             male    M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano    M   36   
2  New Smyrna Beach, Volusia County  Surfing             male    M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella    M  NaN   
4                       Bells Beach  Surfing             male    M  NaN   

                                  Injury Fatal (Y/N)   Time   Species   \
0                  Minor injury to thigh           N  13h00        NaN   
1                   Lacerations to hands           N  11h00        NaN   
2               Lacerations to lower leg           N  10h43        NaN   
3           Struck by fin on chest & leg           N    NaN        NaN   
4  No injury: Knocked off board by shark           N    NaN  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                      href formula  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

                                                                              href  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Case Number.1 Case Number.2  original order Unnamed: 22 Unnamed: 23  
0  2016.09.18.c  2016.09.18.c            5993         NaN         NaN  
1  2016.09.18.b  2016.09.18.b            5992         NaN         NaN  
2  2016.09.18.a  2016.09.18.a            5991         NaN         NaN  
3    2016.09.17    2016.09.17            5990         NaN         NaN  
4    2016.09.16    2016.09.15            5989         NaN         NaN

In [454]:
data.describe().T

count         mean          std  min      25%     50%  \
Year            5992.0  1925.204606   286.473712  0.0  1942.00  1975.0   
original order  5992.0  2997.548899  1729.860210  2.0  1499.75  2997.5   

                    75%     max  
Year            2003.00  2016.0  
original order  4495.25  5993.0

In [455]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 24 columns):
Case Number               5992 non-null object
Date                      5992 non-null object
Year                      5992 non-null int64
Type                      5992 non-null object
Country                   5949 non-null object
Area                      5590 non-null object
Location                  5496 non-null object
Activity                  5465 non-null object
Name                      5792 non-null object
Sex                       5425 non-null object
Age                       3311 non-null object
Injury                    5965 non-null object
Fatal (Y/N)               5973 non-null object
Time                      2779 non-null object
Species                   3058 non-null object
Investigator or Source    5977 non-null object
pdf                       5992 non-null object
href formula              5991 non-null object
href                      5989 non-null object
C

In [456]:
data.shape

(5992, 24)

In [457]:
data.dtypes

Case Number               object
Date                      object
Year                       int64
Type                      object
Country                   object
Area                      object
Location                  object
Activity                  object
Name                      object
Sex                       object
Age                       object
Injury                    object
Fatal (Y/N)               object
Time                      object
Species                   object
Investigator or Source    object
pdf                       object
href formula              object
href                      object
Case Number.1             object
Case Number.2             object
original order             int64
Unnamed: 22               object
Unnamed: 23               object
dtype: object

In [458]:
#data.profile_report()

In [459]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

**The column Species and Sex have spaces at the end on the name, we are going to take them out**

In [460]:
data.columns = ['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23']

# 2. CLEANING THE DATA

## 2.1. We start with the columns

### 2.2.1. Set the index

**The column Original Order 
seems to be the old index. As we will keep  the index of the dataframe, we will delete it**

In [461]:
data = data.drop(['original order'], axis=1)

### 2.2.2. Check if there are empty columns

**Checking the columns with more than 2000 nulls out of 5992**

In [462]:
null_cols = data.isnull().sum()
null_cols[null_cols > 2000]

Age            2681
Time           3213
Species        2934
Unnamed: 22    5991
Unnamed: 23    5990
dtype: int64

**The cols 'Unnamed: 22' and 'Unnamed: 23' have only one value and two values.
We check the values.**

In [463]:
unnamed22 = data['Unnamed: 22'].unique()
unnamed22

array([nan, 'stopped here'], dtype=object)

In [464]:
unnamed23 = data['Unnamed: 23'].unique()
unnamed23

array([nan, 'Teramo', 'change filename'], dtype=object)

**We check the values of the rows of those values.**

In [465]:
data[data['Unnamed: 22']=='stopped here']

Case Number       Date  Year        Type Country        Area  \
1247  2006.06.18  18-Jun-06  2006  Unprovoked  BRAZIL  Pernambuco   

                            Location       Activity                     Name  \
1247  Punta Del Chifre Beach, Olinda  Body boarding  Humberto Pessoa Batista   

     Sex Age                   Injury Fatal (Y/N)   Time Species  \
1247   M  27  Left thigh bitten FATAL           Y  09h00     NaN   

     Investigator or Source                     pdf  \
1247     globalsurfnews.com  2006.06.18-Batista.pdf   

                                                                      href formula  \
1247  http://sharkattackfile.net/spreadsheets/pdf_directory/2006.06.18-Batista.pdf   

                                                                              href  \
1247  http://sharkattackfile.net/spreadsheets/pdf_directory/2006.06.18-Batista.pdf   

     Case Number.1 Case Number.2   Unnamed: 22 Unnamed: 23  
1247    2006.06.18    2006.06.18  stopped here         NaN

In [466]:
data[data['Unnamed: 23']=='Teramo']

Case Number       Date  Year      Type Country        Area  \
4182  1952.07.13  13-Jul-52  1952  Provoked     USA  California   

                         Location Activity  \
4182  San Diego, San Diego County  Fishing   

                                                     Name Sex Age  \
4182  Gerald Howard, on board sportsfishing boat Teresa A   M  34   

                                                             Injury  \
4182  Part of hand removed by shark he had caught PROVOKED INCIDENT   

     Fatal (Y/N) Time Species  Investigator or Source                    pdf  \
4182           N  NaN     NaN  L.A. Times, 7/14/1952   1952.07.13-Howard.pdf   

                                                                     href formula  \
4182  http://sharkattackfile.net/spreadsheets/pdf_directory/1952.07.13-Howard.pdf   

                                                                             href  \
4182  http://sharkattackfile.net/spreadsheets/pdf_directory/1952.07.13-Howard.pdf   

     Case Number.1 Case Number.2 Unnamed: 22 Unnamed: 23  
4182    1952.07.13    1952.07.13         NaN      Teramo

In [467]:
data[data['Unnamed: 23']=='change filename']

Case Number         Date  Year     Type    Country             Area  \
5585  1879.03.10  10-Mar-1879  1879  Invalid  AUSTRALIA  New South Wales   

          Location                                       Activity  \
5585  Near Sydney   The steamship Bonnie Dundee lost in collision   

                                Name Sex  Age  \
5585  Cabin boy of the Bonnie Dundee   M  NaN   

                               Injury Fatal (Y/N) Time Species  \
5585  Partial  remains found in shark           Y  NaN     NaN   

     Investigator or Source                           pdf  \
5585        Star, 3/22/1879  1879.03.10-Bonnie-Dundee.pdf   

                                                                            href formula  \
5585  http://sharkattackfile.net/spreadsheets/pdf_directory/1879.03.10-Bonnie-Dundee.pdf   

                                                                                    href  \
5585  http://sharkattackfile.net/spreadsheets/pdf_directory/1879.03.10-Bonnie-Dundee.pdf   

     Case Number.1 Case Number.2 Unnamed: 22      Unnamed: 23  
5585    1879.03.10    1879.03.10         NaN  change filename

**As the info in the cell seems no relevant, so we drop the columns 'Unnamed: 22' and 'Unnamed: 23'.**

In [468]:
data = data.drop(['Unnamed: 22', 'Unnamed: 23'], axis=1)

In [469]:
data.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  NaN   
4                       Bells Beach  Surfing             male   M  NaN   

                                  Injury Fatal (Y/N)   Time    Species  \
0                  Minor injury to thigh           N  13h00        NaN   
1                   Lacerations to hands           N  11h00        NaN   
2               Lacerations to lower leg           N  10h43        NaN   
3           Struck by fin on chest & leg           N    NaN        NaN   
4  No injury: Knocked off board by shark           N    NaN  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                      href formula  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

                                                                              href  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Case Number.1 Case Number.2  
0  2016.09.18.c  2016.09.18.c  
1  2016.09.18.b  2016.09.18.b  
2  2016.09.18.a  2016.09.18.a  
3    2016.09.17    2016.09.17  
4    2016.09.16    2016.09.15

### 2.2.3 Check if there are duplicate columns to delete them

**We will check if there is any duplicated comlumn**

In [470]:
#source: https://thispointer.com/how-to-find-drop-duplicate-columns-in-a-dataframe-python-pandas/
def getDuplicateColumns(df):
    duplicateColumnNames = set()
    # Iterate over all the columns in dataframe
    for x in range(df.shape[1]):
        # Select column at xth index.
        col = df.iloc[:, x]
        # Iterate over all the columns in DataFrame from (x+1)th index till end
        for y in range(x + 1, df.shape[1]):
            # Select column at yth index.
            otherCol = df.iloc[:, y]
            # Check if two columns at x 7 y index are equal
            if col.equals(otherCol):
                duplicateColumnNames.add(df.columns.values[y])
 
    return list(duplicateColumnNames)

In [471]:
getDuplicateColumns(data)

[]

**There is no complitely duplicate columns.**

**Let's check if with a function we can found possible very similar columns.**

**Create a function to detect percentage of similitud amog 2 rows**

In [472]:
def compareRows(df, col1, col2):
    eq_row = np.where(df[col1] == df[col2], 1, 0).sum()
    rows = df.shape[0] - df[col1].isnull().sum()
    per_row = round((eq_row / rows) * 100,2)
    return per_row

**Make a table comparing column to column**

In [473]:
compared_cols = pd.DataFrame()
compared_cols['index'] = data.columns
compared_cols = compared_cols.set_index('index')
for col in data.columns:
    column = list()
    for col2 in data.columns:
        column.append(compareRows(data, col, col2))
    compared_cols[col] = column
 

In [474]:
compared_cols

Case Number    Date   Year    Type  Country    Area  \
index                                                                         
Case Number                  100.00    0.05    0.0    0.00      0.0    0.00   
Date                           0.05  100.00    0.0    0.00      0.0    0.00   
Year                           0.00    0.00  100.0    0.00      0.0    0.00   
Type                           0.00    0.00    0.0  100.00      0.0    0.00   
Country                        0.00    0.00    0.0    0.00    100.0    0.00   
Area                           0.00    0.00    0.0    0.00      0.0  100.00   
Location                       0.00    0.00    0.0    0.00      0.0    0.13   
Activity                       0.00    0.00    0.0    0.23      0.0    0.00   
Name                           0.00    0.00    0.0    0.00      0.0    0.00   
Sex                            0.00    0.00    0.0    0.00      0.0    0.00   
Age                            0.00    0.00    0.0    0.00      0.0    0.00   
Injury                         0.00    0.00    0.0    0.00      0.0    0.00   
Fatal (Y/N)                    0.00    0.00    0.0    0.00      0.0    0.00   
Time                           0.00    0.00    0.0    0.00      0.0    0.00   
Species                        0.00    0.00    0.0    0.00      0.0    0.00   
Investigator or Source         0.00    0.00    0.0    0.00      0.0    0.00   
pdf                            0.02    0.00    0.0    0.00      0.0    0.00   
href formula                   0.00    0.00    0.0    0.00      0.0    0.00   
href                           0.00    0.00    0.0    0.00      0.0    0.00   
Case Number.1                 99.78    0.05    0.0    0.00      0.0    0.00   
Case Number.2                 99.97    0.05    0.0    0.00      0.0    0.00   

                        Location  Activity    Name     Sex    Age  Injury  \
index                                                                       
Case Number                 0.00      0.00    0.00    0.00    0.0    0.00   
Date                        0.00      0.00    0.00    0.00    0.0    0.00   
Year                        0.00      0.00    0.00    0.00    0.0    0.00   
Type                        0.00      0.26    0.00    0.00    0.0    0.00   
Country                     0.00      0.00    0.00    0.00    0.0    0.00   
Area                        0.13      0.00    0.00    0.00    0.0    0.00   
Location                  100.00      0.00    0.00    0.00    0.0    0.00   
Activity                    0.00    100.00    0.00    0.02    0.0    0.00   
Name                        0.00      0.00  100.00    0.00    0.0    0.02   
Sex                         0.00      0.02    0.00  100.00    0.0    0.00   
Age                         0.00      0.00    0.00    0.00  100.0    0.00   
Injury                      0.00      0.00    0.02    0.00    0.0  100.00   
Fatal (Y/N)                 0.00      0.00    0.00    0.02    0.0    0.00   
Time                        0.00      0.00    0.00    0.00    0.0    0.02   
Species                     0.00      0.00    0.00    0.00    0.0    0.02   
Investigator or Source      0.00      0.00    0.00    0.00    0.0    0.00   
pdf                         0.00      0.00    0.00    0.00    0.0    0.00   
href formula                0.00      0.00    0.00    0.00    0.0    0.00   
href                        0.00      0.00    0.00    0.00    0.0    0.00   
Case Number.1               0.00      0.00    0.00    0.00    0.0    0.00   
Case Number.2               0.00      0.00    0.00    0.00    0.0    0.00   

                        Fatal (Y/N)    Time  Species  Investigator or Source  \
index                                                                          
Case Number                    0.00    0.00     0.00                     0.0   
Date                           0.00    0.00     0.00                     0.0   
Year                           0.00    0.00     0.00                     0.0   
Type                           0.0

**Case number is very similar to Case Number.1 and Case Sumber.2**
**href and href formula are very similar too.**

**Lets chek the different values is Case number, Case Number.1 and Case Sumber.2**

In [475]:
data[['Date','Year','Sex','Case Number','Case Number.1','Case Number.2']].loc[(data['Case Number'] != data['Case Number.1'])| (data['Case Number'] != data['Case Number.2']) | (data['Case Number.1'] != data['Case Number.2'])]

Date  Year  Sex   Case Number  \
4                                16-Sep-16  2016    M    2016.09.15   
33                   Reported  14-Jul-2016  2016    M  2016.07.14.4   
97                               24-Jan-16  2016    M  2016.01.24.b   
116                              07-Nov-15  2015    F    2015.12.23   
121                              28-Oct-15  2015    M  2015.10.28.a   
169                              10-Jul-15  2015    M    2015.07-10   
3296                             05-Jul-67  1967    M    1967.07.05   
3569                             30-Aug-62  1962    M  1962,08.30.b   
3654                  Reported 06-Sep-1961  1961    M  1961.09.02.R   
4177                             05-Aug-52  1952  NaN    1952.08.05   
4719                  Reported 08-Feb-1934  1934    M  1934.01.08.R   
5043  Reported to have taken place in 1919  1919    M  1900.00.00.R   
5150                  Reported 31-Jul-1911  1911    M  1911.07.31.R   

     Case Number.1 Case Number.2  
4       2016.09.16    2016.09.15  
33    2016.07.14.R  2016.07.14.4  
97    2015.01.24.b  2016.01.24.b  
116     2015.11.07    2015.12.23  
121     2015.10.28  2015.10.28.a  
169     2015.07.10    2015.07.10  
3296    1967/07.05    1967.07.05  
3569  1962.08.30.b  1962,08.30.b  
3654  1961.09,06.R  1961.09.02.R  
4177    1952.08.04    1952.08.05  
4719  1934.02.08.R  1934.02.08.R  
5043  1919.00.00.R  1900.00.00.R  
5150  1911.07.31.T  1911.07.31.R

**First I will make the columns to be type string and then replace all the comas in the Case Number, Case Number.1 and Case Number.2**
**As I guess the rules for the naming are year.month.day.victim.Reported yyyy.mm.dd.[a].[R]**

In [476]:
data[['Case Number', 'Case Number.1', 'Case Number.2']] = data[['Case Number', 'Case Number.1', 'Case Number.2']].astype(str)


In [477]:
data['Case Number']= data['Case Number'].str.replace(',', '.')

In [478]:
data['Case Number.1']= data['Case Number.1'].str.replace(',', '.')

In [479]:
data['Case Number.2']= data['Case Number.2'].str.replace(',', '.')

**I think Case Number.1 and Case Number.2 are corrections about the Case Number**

In [480]:
data[['Date','Year','Sex','Case Number','Case Number.1','Case Number.2']].loc[(data['Case Number'] != data['Case Number.1'])| (data['Case Number'] != data['Case Number.2']) | (data['Case Number.1'] != data['Case Number.2'])]

Date  Year  Sex   Case Number  \
4                                16-Sep-16  2016    M    2016.09.15   
33                   Reported  14-Jul-2016  2016    M  2016.07.14.4   
97                               24-Jan-16  2016    M  2016.01.24.b   
116                              07-Nov-15  2015    F    2015.12.23   
121                              28-Oct-15  2015    M  2015.10.28.a   
169                              10-Jul-15  2015    M    2015.07-10   
3296                             05-Jul-67  1967    M    1967.07.05   
3654                  Reported 06-Sep-1961  1961    M  1961.09.02.R   
4177                             05-Aug-52  1952  NaN    1952.08.05   
4719                  Reported 08-Feb-1934  1934    M  1934.01.08.R   
5043  Reported to have taken place in 1919  1919    M  1900.00.00.R   
5150                  Reported 31-Jul-1911  1911    M  1911.07.31.R   

     Case Number.1 Case Number.2  
4       2016.09.16    2016.09.15  
33    2016.07.14.R  2016.07.14.4  
97    2015.01.24.b  2016.01.24.b  
116     2015.11.07    2015.12.23  
121     2015.10.28  2015.10.28.a  
169     2015.07.10    2015.07.10  
3296    1967/07.05    1967.07.05  
3654  1961.09.06.R  1961.09.02.R  
4177    1952.08.04    1952.08.05  
4719  1934.02.08.R  1934.02.08.R  
5043  1919.00.00.R  1900.00.00.R  
5150  1911.07.31.T  1911.07.31.R

**I pick the correct data from the other columns**

In [481]:
data['Case Number'].iloc[169] = data['Case Number.2'].iloc[169]

In [482]:
data['Case Number'].iloc[[4, 33, 116, 121, 3569, 3654, 4719, 5043]] = data['Case Number.1'].iloc[[4, 33, 116, 121, 3569, 3654, 4719, 5043]]


In [483]:
data[['Date','Year','Sex','Case Number','Case Number.1','Case Number.2']].loc[(data['Case Number'] != data['Case Number.1'])| (data['Case Number'] != data['Case Number.2']) | (data['Case Number.1'] != data['Case Number.2'])]

Date  Year  Sex   Case Number  \
4                                16-Sep-16  2016    M    2016.09.16   
33                   Reported  14-Jul-2016  2016    M  2016.07.14.R   
97                               24-Jan-16  2016    M  2016.01.24.b   
116                              07-Nov-15  2015    F    2015.11.07   
121                              28-Oct-15  2015    M    2015.10.28   
3296                             05-Jul-67  1967    M    1967.07.05   
3654                  Reported 06-Sep-1961  1961    M  1961.09.06.R   
4177                             05-Aug-52  1952  NaN    1952.08.05   
5043  Reported to have taken place in 1919  1919    M  1919.00.00.R   
5150                  Reported 31-Jul-1911  1911    M  1911.07.31.R   

     Case Number.1 Case Number.2  
4       2016.09.16    2016.09.15  
33    2016.07.14.R  2016.07.14.4  
97    2015.01.24.b  2016.01.24.b  
116     2015.11.07    2015.12.23  
121     2015.10.28  2015.10.28.a  
3296    1967/07.05    1967.07.05  
3654  1961.09.06.R  1961.09.02.R  
4177    1952.08.04    1952.08.05  
5043  1919.00.00.R  1900.00.00.R  
5150  1911.07.31.T  1911.07.31.R

**Now I can drop the columns Case Number.1 and Case Number.2**

In [484]:
data = data.drop(['Case Number.1', 'Case Number.2'], axis=1)

In [485]:
data.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  NaN   
4                       Bells Beach  Surfing             male   M  NaN   

                                  Injury Fatal (Y/N)   Time    Species  \
0                  Minor injury to thigh           N  13h00        NaN   
1                   Lacerations to hands           N  11h00        NaN   
2               Lacerations to lower leg           N  10h43        NaN   
3           Struck by fin on chest & leg           N    NaN        NaN   
4  No injury: Knocked off board by shark           N    NaN  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                      href formula  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

                                                                              href  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

**Lets chek the different values href and href formula**

In [486]:
data[['Date','Year','Sex','href','href formula','Case Number']].loc[data['href'] != data['href formula']]


Date  Year  Sex  \
20               29-Jul-16  2016    M   
27               23-Jul-16  2016    M   
61               21-May-16  2016    F   
107              21-Dec-15  2015    M   
114              15-Nov-15  2015    M   
134              05-Oct-15  2015    M   
180              27-Jun-15  2015    M   
193              13-Jun-15  2015    F   
232              24-Jan-15  2015    M   
262              19-Oct-14  2014    F   
263              18-Oct-14  2014    M   
264              17-Oct-14  2014    F   
271              03-Oct-14  2014    M   
272              03-Oct-14  2014    M   
293              12-Aug-14  2014    M   
305              22-Jul-14  2014    M   
323              09-Jun-14  2014    M   
347              03-Apr-14  2014    F   
361              07-Feb-14  2014    M   
362              07-Feb-14  2014    M   
363              26-Jan-14  2014    M   
364              25-Jan-14  2014    M   
365              04-Jan-14  2014    M   
367              25-Dec-13  2013    M   
377              22-Nov-13  2013    M   
378              12-Nov-13  2013    M   
379              10-Nov-13  2013    M   
380              07-Nov-13  2013    M   
381              07-Nov-13  2013    M   
382              31-Oct-13  2013    M   
383              28-Oct-13  2013    F   
384              26-Oct-13  2013    M   
448              27-May-13  2013    M   
449              27-May-13  2013    M   
823              01-Feb-10  2010    F   
1217             22-Aug-06  2006    M   
1218             22-Aug-06  2006    M   
2274                Apr-93  1993    M   
2477                  1989  1989    M   
3019             19-Jan-75  1975    M   
3549                  1963  1963    M   
3603             25-Mar-62  1962    M   
3917             31-Aug-58  1958  NaN   
3928                Jul-58  1958    M   
4394             30-Jul-45  1945  NaN   
4642  Reported 04-Aug-1936  1936    M   
4668             24-Aug-35  1935    M   
4719  Reported 08-Feb-1934  1934    M   
5317             05-Sep-00  1900    M   
5458            06-26-1890  1890    M   
5686           02-Jan-1864  1864    M   
5694         Reported 1863  1863    M   
5819  Reported 08-Jul-1819  1819    M   
5857         Reported 1637  1637  NaN   

                                                                                                                                         href  \
20           http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2016.07.29-Spain.pdf   
27                                                              http://sharkattackfile.net/spreadsheets/pdf_directory/2016.07.23-Cutbirth.pdf   
61          http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2016.05.21.a-Girl.pdf   
107       http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2015.12.21.a-Brazil.pdf   
114     http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2015.11.15.a-Engelman.pdf   
134                                                            http://sharkattackfile.net/spreadsheets/pdf_directory/2015.10.05-Slaughter.pdf   
180                                                            http://sharkattackfile.net/spreadsheets/pdf_directory/2015.06.27-Swanepoel.pdf   
193                                                                                                                                       NaN   
232         http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2015.01.24-Murray.pdf   
262           http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.19-Tara.pdf   
263                                                                 http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.18-Roberson   
264                       

**I assume href formula is the most accurate column for the urls, anyhow I will correct some errors**

**Take some values from href as the cell has the correct url**

In [487]:
data['href formula'].iloc[[193,293,364,448,823,3019]] = data['href'].iloc[[193,293,364,448,823,3019]]

In [488]:
data['href formula'].iloc[363] = None

In [489]:
index_href = data['href formula'].loc[data['href formula'].str.endswith('.pdf') == False].index

In [490]:
data['href formula'].iloc[index_href]

65               http://sharkattackfile.net/spreadsheets/pdf_directory/2016.05.02-NZ-pdf
126      http://sharkattackfile.net/spreadsheets/pdf_directory/2015.10.17.b.-Hawaii. pdf
157           http://sharkattackfile.net/spreadsheets/pdf_directory/2015.08.19-Szsarmack
177          http://sharkattackfile.net/spreadsheets/pdf_directory/2015.06.30.b-Steve.pf
226         http://sharkattackfile.net/spreadsheets/pdf_directory/2015.02.14-Bishop.pdf 
251      http://sharkattackfile.net/spreadsheets/pdf_directory/2014.11.20-Mauritius.pdf 
252           http://sharkattackfile.net/spreadsheets/pdf_directory/2014.11.19-Freo.pdf 
264         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.17-Olsson, pdf
295      http://sharkattackfile.net/spreadsheets/pdf_directory/2014.08.09.R-Hindley, pdf
334         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.05.14-McLeod, pdf
343       http://sharkattackfile.net/spreadsheets/pdf_directory/.2014.04.12.R-inflatable
352            http:/

In [491]:
data['href formula'] = data['href formula'].str.replace(', pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace('. Pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace('-pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace('.pf', '.pdf')
data['href formula'] = data['href formula'].str.replace('. pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace(' ','')
data['href formula'] = data['href formula'].str.replace('/ .','')

In [492]:
data['href formula'] = data['href formula'].str.replace('/.2014','/2014')

In [493]:
data['href formula'].loc[data['href formula'].str.endswith('.pdf') == False]

157        http://sharkattackfile.net/spreadsheets/pdf_directory/2015.08.19-Szsarmack
343     http://sharkattackfile.net/spreadsheets/pdf_directory/2014.04.12.R-inflatable
352         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.03.18.c-Borgna
359         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.02.17.R-OneDLL
425       http://sharkattackfile.net/spreadsheets/pdf_directory/2013.07.29.a-Gatewood
1004          http://sharkattackfile.net/spreadsheets/pdf_directory/2008.07.26.a-Maui
2476      http://sharkattackfile.net/spreadsheets/pdf_directory/1989.00.00.b-Halstead
2714        http://sharkattackfile.net/spreadsheets/pdf_directory/1983.06.00-Starling
3987          http://sharkattackfile.net/spreadsheets/pdf_directory/1957.00.00.i-baby
4481       http://sharkattackfile.net/spreadsheets/pdf_directory/1942.07.06-Bardolini
5257                 http://sharkattackfile.net/spreadsheets/pdf_directory/1905.07.00
5329         http://sharkattackfile.net/spreadsheets/p

**The href formula is as complete and accurate I can. I'm going to delete the href column and rename the href formula column**

In [494]:
data = data.drop(['href'], axis=1)

In [495]:
data = data.rename(index=str, columns={"href formula": "url"})

In [496]:
data.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  NaN   
4                       Bells Beach  Surfing             male   M  NaN   

                                  Injury Fatal (Y/N)   Time    Species  \
0                  Minor injury to thigh           N  13h00        NaN   
1                   Lacerations to hands           N  11h00        NaN   
2               Lacerations to lower leg           N  10h43        NaN   
3           Struck by fin on chest & leg           N    NaN        NaN   
4  No injury: Knocked off board by shark           N    NaN  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                               url  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

### 2.2.3 Columns with a lot of Null Values

**Three more columns have a lot of Nulls, we check them: Age, Time, Species.**

**Lests check the different values of Age**

In [497]:
data['Age'].unique()

array(['16', '36', '43', nan, '60s', '51', '50', '12', '9', '22', '25',
       '37', '20', '49', '15', '21', '40', '72', '18', '29', '31', '11',
       '10', '59', '42', '34', '35', '19', '6', '27', '64', '60', '23',
       '52', '13', '57', '48', '39', '24', '26', '69', '46', 'Teen', '41',
       '45', '65', '38', '71', '32', '58', '28', '54', '44', '14', '7',
       '62', '40s', '68', '47', '17', '30', '63', '70', '18 months', '53',
       '20s', '33', '30s', '50s', '8', '61', '55', 'teen', '66', '77',
       '74', '3', '56', '28 & 26', '5', '86', '18 or 20', '12 or 13',
       '46 & 34', '28, 23 & 30', 'Teens', '36 & 26', '8 or 10', '84',
       '\xa0 ', ' ', '30 or 36', '6½', '21 & ?', '75', '33 or 37',
       'mid-30s', '73', '23 & 20', '7      &    31', '20?', "60's",
       '32 & 30', '16 to 18', '87', '67', 'Elderly', 'mid-20s', 'Ca. 33',
       '21 or 26', '>50', '18 to 22', 'adult', '9 & 12', '? & 19',
       '9 months', '25 to 35', '23 & 26', '1', '(adult)', '33 & 37',
     

**I will change all the non numeric values by NaN, pass the Age to numeric, fill the NaN to 0.0 and finally convert all the types to integer**

In [498]:
data['Age'] = pd.to_numeric(data['Age'], errors='coerce') 

In [499]:
data['Age'] = data['Age'].fillna(-1)

In [500]:
data['Age'] = data['Age'].astype(int)

In [501]:
data['Age'].unique()

array([16, 36, 43, -1, 51, 50, 12,  9, 22, 25, 37, 20, 49, 15, 21, 40, 72,
       18, 29, 31, 11, 10, 59, 42, 34, 35, 19,  6, 27, 64, 60, 23, 52, 13,
       57, 48, 39, 24, 26, 69, 46, 41, 45, 65, 38, 71, 32, 58, 28, 54, 44,
       14,  7, 62, 68, 47, 17, 30, 63, 70, 53, 33,  8, 61, 55, 66, 77, 74,
        3, 56,  5, 86, 84, 75, 73, 87, 67,  1, 81, 78])

**Lests check the different values of Time**

In [502]:
data['Time'].unique()

array(['13h00', '11h00', '10h43', nan, '15h15', '14h30', '15h40',
       'Late afternoon', '15h00', '14h00', '17h00', '16h00', '11h30',
       '12h00', '19h05', 'After noon', '1300', '10h00', '14h30 / 15h30',
       '22h00', '16h20', '14h34', '15h25', '14h55', '06h00', '17h30',
       '08h30', 'Morning', 'Afternoon', '15h45', '18h00', '17h46',
       'Morning ', '13h20', '15h49', '07h00', '08h00', '10h45', '19h00',
       '13h30', 'Midnight', '12h30', '09h30 / 10h00', '10h30', '18h15',
       '04h00', '14h50', '13h50', '19h20', '10h25', '16h50',
       '10h45-11h15', '16h45', '15h52', '07h30', '06h15', '14h40',
       '19h45', '17h10', '15h30', '10h10', '08h45', 'Evening', '09h00',
       '12h10', '18h05', '19h30', '11h41', '12h25', '09h30', '17h51',
       '16h12', '12h45', '10h50', '13h45', '14h15', '16h30', '09h45',
       '11h45', '10h20', '18h30', '06h45', '05h00', '03h30', 'Night',
       'Sometime between 06h00 & 08hoo', '06h30', '11h15', '17h20',
       '18h45', '20h00', '16h40

In [503]:
data['Time'] = data['Time'].str.replace('h','')
data['Time'] = data['Time'].str.replace('j','')

In [504]:
data['Time'].unique()

array(['1300', '1100', '1043', nan, '1515', '1430', '1540',
       'Late afternoon', '1500', '1400', '1700', '1600', '1130', '1200',
       '1905', 'After noon', '1000', '1430 / 1530', '2200', '1620',
       '1434', '1525', '1455', '0600', '1730', '0830', 'Morning',
       'Afternoon', '1545', '1800', '1746', 'Morning ', '1320', '1549',
       '0700', '0800', '1045', '1900', '1330', 'Midnigt', '1230',
       '0930 / 1000', '1030', '1815', '0400', '1450', '1350', '1920',
       '1025', '1650', '1045-1115', '1645', '1552', '0730', '0615',
       '1440', '1945', '1710', '1530', '1010', '0845', 'Evening', '0900',
       '1210', '1805', '1930', '1141', '1225', '0930', '1751', '1612',
       '1245', '1050', '1345', '1415', '1630', '0945', '1145', '1020',
       '1830', '0645', '0500', '0330', 'Nigt',
       'Sometime between 0600 & 08oo', '0630', '1115', '1720', '1845',
       '2000', '1640', '1618', '1110', 'Early afternoon', '1750',
       '0700 - 0800', '1815-1830', '1701', '1015', '0957'

In [505]:
data['Time'] = pd.to_numeric(data['Time'], errors='coerce') 

In [506]:
data['Time'] = data['Time'].fillna(-1.0)

In [507]:
data['Time'] = data['Time'].astype(int)

In [508]:
data['Time'].unique()

array([ 1300,  1100,  1043,    -1,  1515,  1430,  1540,  1500,  1400,
        1700,  1600,  1130,  1200,  1905,  1000,  2200,  1620,  1434,
        1525,  1455,   600,  1730,   830,  1545,  1800,  1746,  1320,
        1549,   700,   800,  1045,  1900,  1330,  1230,  1030,  1815,
         400,  1450,  1350,  1920,  1025,  1650,  1645,  1552,   730,
         615,  1440,  1945,  1710,  1530,  1010,   845,   900,  1210,
        1805,  1930,  1141,  1225,   930,  1751,  1612,  1245,  1050,
        1345,  1415,  1630,   945,  1145,  1020,  1830,   645,   500,
         330,   630,  1115,  1720,  1845,  2000,  1640,  1618,  1110,
        1750,  1701,  1015,   957,   820,  1758,  2030,  1519,  1055,
        1555,  2300,  1240,  1355,  1605,  1410,  1520,  1250,  1445,
        1340,  1324,   940,  1140,   810,  1556,  1420,   756,  1717,
        1635,   905,   715,   720,  1928,  1238,   550,  1550,  1105,
         545,  1325,   950,  1220,  1326,   911,  1215,  1315,  1745,
        1820,  1040,

In [509]:
data['hour'] = round(data['Time']/100).astype(int)

In [510]:
data['hour'].loc[data['hour'] > 24] = -1

In [511]:
data['hour'].unique()

array([13, 11, 10,  0, 15, 14, 17, 16, 12, 19, 22,  6,  8, 18,  7,  4,  9,
        5,  3, 20, 23, 21, -1,  2,  1])

In [516]:
data['min'] = round(data['Time']%100).astype(int)

In [517]:
data['min'].loc[data['hour'] > 59] = -1

In [532]:
data['min'].loc[data['min'] == 0] = '00'
data['min'].loc[data['min'] == 1] = '01'
data['min'].loc[data['min'] == 2] = '02'
data['min'].loc[data['min'] == 3] = '03'
data['min'].loc[data['min'] == 4] = '04'
data['min'].loc[data['min'] == 5] = '05'
data['min'].loc[data['min'] == 6] = '06'
data['min'].loc[data['min'] == 7] = '07'
data['min'].loc[data['min'] == 8] = '08'
data['min'].loc[data['min'] == 9] = '09'
data['min'].loc[data['hour'] == -1] = ''

In [533]:
data['min'].unique()

array(['00', 43, 99, 15, 30, 40, '05', 20, 34, 25, 55, 45, 46, 49, 50, 52,
       10, 41, 51, 12, 18, '01', 57, 58, 19, 24, 56, 17, 35, 28, 38, 26,
       11, 44, 27, 16, 21, 53, '', 42, 39, '08', '02', 14, 37, '07', 23,
       47, '06', 48, 33, 54, 22, 32], dtype=object)

In [534]:
data['Time'] = data['hour'].astype(str) + ':' + data['min'].astype(str) 

In [535]:
data['Time'].loc[data['hour'] == -1] = ''

In [536]:
data['Time'].unique()

array(['13:00', '11:00', '10:43', '0:99', '15:15', '14:30', '15:40',
       '15:00', '14:00', '17:00', '16:00', '11:30', '12:00', '19:05',
       '10:00', '22:00', '16:20', '14:34', '15:25', '15:55', '6:00',
       '17:30', '8:30', '15:45', '18:00', '17:46', '13:20', '15:49',
       '7:00', '8:00', '10:45', '19:00', '13:30', '12:30', '10:30',
       '18:15', '4:00', '14:50', '19:20', '10:25', '16:50', '16:45',
       '16:52', '7:30', '6:15', '14:40', '19:45', '17:10', '15:30',
       '10:10', '8:45', '9:00', '12:10', '18:05', '19:30', '11:41',
       '12:25', '9:30', '18:51', '16:12', '12:45', '10:50', '13:45',
       '14:15', '16:30', '9:45', '11:45', '10:20', '18:30', '6:45',
       '5:00', '3:30', '6:30', '11:15', '17:20', '18:45', '20:00',
       '16:40', '16:18', '11:10', '18:50', '17:01', '10:15', '10:57',
       '8:20', '18:58', '20:30', '15:19', '11:55', '16:55', '23:00',
       '12:40', '14:55', '16:05', '14:10', '15:20', '12:50', '14:45',
       '13:40', '13:24', '9:40', '11:

In [537]:
data = data.drop(['min'], axis=1)

In [542]:
data.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M   -1   
4                       Bells Beach  Surfing             male   M   -1   

                                  Injury Fatal (Y/N)   Time    Species  \
0                  Minor injury to thigh           N  13:00        NaN   
1                   Lacerations to hands           N  11:00        NaN   
2               Lacerations to lower leg           N  10:43        NaN   
3           Struck by fin on chest & leg           N   0:99        NaN   
4  No injury: Knocked off board by shark           N   0:99  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

   hour  
0    13  
1    11  
2    10  
3     0  
4     0

**Lests check the different values of species**

In [543]:
species = data['Species'].unique()

In [545]:
for specie in species:
    print(specie)

nan
2 m shark
3' to 4' shark
Tiger shark, 10?
White shark
Bull shark, 6'
Bull shark, 3.5 m
Tiger shark, 10' 
Blacktip shark
Blue shark
Nurse shark, 4'
reef shark, 1m
Lemon shark, 9'
Angel shark
dogfish shark
Tawny nurse shark
Caribbean reef shark, 4.5'
3' to 5' shark
5' shark
3' shark
Said to involve an 8' shark but more likely damage caused by debris
3+ m shark
Mako shark
Bronze whaler sharks x 3
3 m shark
Bull shark, 4' to 5'
Blacktip shark, 4'
Nurse shark, 2'
Mako shark, 1.5 m [5'] 
Wobbegong shark
Grey reef shark, 2 m
Bull shark, 7'
Tiger shark, 2.5 m
Shark involvement not confirmed
5' to 6' shark
Nurse shark, 2.5-ft
Bronze whaler
Sandtiger shark
No shark involvement
Reef shark, 5'
Bull sharks x 2
Tiger shark, 14'
White shark, 3 m
Silky shark, 6.5'
9' shark
Bronze whaler shark, 6'
Tiger shark, 1.5 m 
White shark, 3.5 m
Bull shark
Spinner shark, 7'
Bull shark, 2.8 to 3.1 m
White shark, 3 to 3.5m 
Spinner shark, 5'
Tiger shark, 7'
Hammerhead sp.
Tiger shark, 13' 
4' to 5' shark
Black

**Let's Work with the column YEAR**

In [546]:
data['Year'].unique()

array([2016, 2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007, 2006,
       2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998, 1997, 1996, 1995,
       1984, 1994, 1993, 1992, 1991, 1990, 1989, 1969, 1988, 1987, 1986,
       1985, 1983, 1982, 1981, 1980, 1979, 1978, 1977, 1976, 1975, 1974,
       1973, 1972, 1971, 1970, 1968, 1967, 1966, 1965, 1964, 1963, 1962,
       1961, 1960, 1959, 1958, 1957, 1956, 1955, 1954, 1953, 1952, 1951,
       1950, 1949, 1948, 1848, 1947, 1946, 1945, 1944, 1943, 1942, 1941,
       1940, 1939, 1938, 1937, 1936, 1935, 1934, 1933, 1932, 1931, 1930,
       1929, 1928, 1927, 1926, 1925, 1924, 1923, 1922, 1921, 1920, 1919,
       1918, 1917, 1916, 1915, 1914, 1913, 1912, 1911, 1910, 1909, 1908,
       1907, 1906, 1905, 1904, 1903, 1902, 1901, 1900, 1899, 1898, 1897,
       1896, 1895, 1894, 1893, 1892, 1891, 1890, 1889, 1888, 1887, 1886,
       1885, 1884, 1883, 1882, 1881, 1880, 1879, 1878, 1877, 1876, 1875,
       1874, 1873, 1872, 1871, 1870, 1869, 1868, 18

**It seems there are some outliers. Lets Check it**

In [ ]:
data.loc[(data['Year'] == 0)]

## 2.2. We continue with the rows


### Check the duplicates rows

In [ ]:
duplicate_rows = data[data.duplicated(['Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species', 'Investigator or Source'])]

In [ ]:
duplicate_rows

In [ ]:
data.loc[(data['Country'] == 'USA') & (data['Area'] == 'Hawaii') 
         & (data['Year'] == 1943) & (data['Activity'] == 'Spearfishing')]

We see that the row has "2 males" in name, so that is why the row is probably duplicate.

We won't delet the rows.